### DISCUSSION TITLE
Spacy Pretraining

### DISCUSSION THREAD

**Discussion by sneha5gsm** [explosion/spaCy#7999] [Help: Coding & Implementations] [answered]
Hi
I wanted to understand what Spacy's pretrain command does. I am trying to train an NER model.
From whatever I read in the Spacy docs, we can initialize inputs in the following ways:
1. Use vectors found online or trained separately and convert it to Spacy format. Make changes in the model config to include those vectors in the  'vectors' path section.
2. Use spacy's trained vectors (web_en_core_lg, etc) and we won't need to initialize anything in the model config file.
3. Use tok2vec pipeline (or other word embedding algorithm pipeline) and use Spacy's pretrain command to train vectors from scratch. Then use the model it in the 'init_tok2vec' path section of the model config

What worked best : Train vectors separately and format them so that spacy can use it and use it in the 'vectors' path section of the model and pretraining model config files. Use spacy's pretrain command and use the model it in the 'init_tok2vec' path section of the model config.

Questions:
1. Are there more ways to initialize vectors for a model? Are any of the above mentioned ways wrong?
2. I was thinking only one of the arguments among 'vectors', 'init_tok2vec' was enough to initialize vectors for the model. But initializing both worked better, why? What happens in this scenario? For our pretraining model config, we used the 'vectors' argument (vectors trained separately). What does this mean? Does the pretraining model initialize the inputs using these vectors and start training from there?

Thanks!

**Comment by polm** [comment, marked as answer]
> Are there more way to initialize vectors for a model?

No.

> Are any of the above mentioned ways wrong?

tok2vec and static vectors are different things. spaCy does not have any feature to train vectors from scratch. It's for a former version of spaCy, but @svlandeg has a great explanation of the relationship between tok2vec and static vectors [here](https://stackoverflow.com/questions/63144230/proper-way-to-add-new-vectors-for-oov-words/63520262#63520262).

> But initializing both worked better, why? What happens in this scenario?

When you specify vectors and use them in pretraining, your tok2vec layer can learn from both your raw corpus and the information encoded in the vectors. Then when you use this tok2vec in your downstream task it can use information from all these different sources. There's nothing models like as much as extra information, and all that delicious data helps your model perform better.

> For our pretraining model config, we used the 'vectors' argument (vectors trained separately). What does this mean? Does the pretraining model initialize the inputs using these vectors and start training from there?

This affects the pretraining model objective, you can read about that [here](https://spacy.io/usage/embeddings-transformers#:~:text=the%20%22vectors%22%20objective%20asks%20the%20model%20to%20predict%20the%20word%E2%80%99s%20vector). Basically it just makes the tok2vec layer learn to predict the vectors in addition to the other tasks it's doing, which lets it learn from the vectors.

**Reply by sneha5gsm** [reply]
Thanks! This helps.

I guess it's confusing to understand this in the [docs](https://spacy.io/usage/embeddings-transformers#pretraining), esp this part:
'The spacy pretrain command lets you initialize your models with information from raw text. Without pretraining, the models for your components will usually be initialized randomly. The idea behind pretraining is simple: random probably isn’t optimal, so if we have some text to learn from, we can probably find a way to get the model off to a better start.'

I guess I don't entirely understand how the weights from the pretrained model will help initialize the weights of other components in the model (like NER component). Won't it just initialize the pretrain (toke2vec) component?

My current understanding from the docs and what you said is:
Spacy won't create/train vectors from scratch using the pretraining command. We would need to use other libraries or trained embeddings or spacy's trained embeddings (en_core_web_lg, etc) to initialize inputs for our models. Pretraining involves adding another component to our model. The component learns stuff from our training data (converted into numbers using our embeddings(vectors)) and later on uses it when we include this component in our model pipeline (like NER, etc). But the weights learned for the pretraining components will only be useful to initialize that component, not other pipelines. The pretraining component basically got early access to our data than the rest of our model and it might help our model learn better by giving more information about our data.

1. Is my understanding correct?
2. The linked [article](https://stackoverflow.com/questions/63144230/proper-way-to-add-new-vectors-for-oov-words/63520262#63520262) on tok2vec is very helpful. What does 'token to vector (tok2vec)' mean? (Like how Word2Vec means it converts words to vectors.) Are the tokens referring to the input vectors? 
3. Does Spacy's [trained pipelines](https://spacy.io/models/en) like en_core_web_lg have trained vectors for initialization? How were those vectors created? Are they trained from scratch on huge data corpuses using word embedding algorithms.
4. What happens when we use en_core_web_lg, etc in the pretrain command. Are the inputs initialized with those vectors and further trained?
5. My understanding is the 'vectors' in the path section of the config is for initializing our text input to numbers. What happens when we don't provide the 'vectors' in the path section of the config, how are the inputs initialized? Are the inputs one hot encoded? It says in the docs that 'vectors' in the config defaults to null.  Also, the terminology of using 'vectors' here and in 'tok2vec' is confusing.

Thanks!

**Reply by polm** [reply]
I think your understanding is still a little off. 

Static vectors and tok2vec both take an input word (or token) and return a vector, which is just a list of floating point numbers. But the way they do this is a bit different.

**Static vectors**, like those trained by word2vec, are like a Python dictionary. If you get the vector for "bank" it will be the same no matter what sentence it is in. The way these are made varies by method, but for word2vec and most methods they start out random, and they are adjusted so that words that appear in similar surroundings are more numerically similar (= closer in vector space). 

**tok2vec** is more like a *function*, in that context affects the output. So you can think of it like `tok2vec("bank", context)`, and if the context (the surrounding words) change, then the returned vector will change. word2vec also uses context, but only at training time, while tok2vec uses context at inference time too. 

Another thing that makes spaCy's tok2vec different from word2vec is that it takes into account a variety of features of the word. word2vec only considers the literal "surface" or text form of the word, while spaCy uses features like prefix, suffix, shape, and so on. 

-----

To answer your other questions...

2. "token to vector" means converting a token to a vector. A "token" is the unit of output of the tokenizer, it's usually the same as a word. (Example of not being the same: "don't" is one word but two tokens, "do" and "n't", to make it more consistent with the unabbreviated form "do not".)

3. All of spaCy's pipelines have tok2vec layers. The medium and large models include static vectors. The way the vectors were created depends on language, though I think all of them were created by other projects and we re-use them; for example, for the English models the vectors are from GloVe. (This is accidentally missing from the docs now, but will be put back in for the next release.)

4. You can use existing models for their static vectors, which would use those as input for training a randomly initialized tok2vec, or you can start with the pretrained tok2vec and continue training it. 

5. If you don't have vectors in a model then tok2vec is just trained on inputs other than vectors; vectors are only one of the possible training objectives. There's no helping that both "static vectors" and "tok2vec" use the term vectors because that's what their output is.

**Reply by sneha5gsm** [reply]
Thanks for the quick reply!

Since the tokens are the inputs for tok2vec, how are the tokens(inputs) initialized in tok2vec? or How are the tokens converted into numbers for training tok2vec? Are tokens one hot vectors? 
If we give the 'vectors' in the path section of the pretraining(tok2vec) config file, what's the use. What is it used for?

> If you don't have vectors in a model then tok2vec is just trained on inputs other than vectors; vectors are only one of the possible training objectives. 

But it's better to provide vectors?

Thanks

**Reply by polm** [reply]
> Since the tokens are the inputs for tok2vec, how are the tokens(inputs) initialized in tok2vec? or How are the tokens converted into numbers for training tok2vec? Are tokens one hot vectors?

They are not one hot vectors, the strings are converted to vectors using HashEmbed. You can read more about HashEmbed [here](https://thinc.ai/docs/api-layers#hashembed) and [here](https://support.prodi.gy/t/can-you-explain-how-exactly-hashembed-works/564).

> But it's better to provide vectors?

It depends on the amount and type of data you have. If in doubt I recommend you compare results with and without vectors.

**Reply by sneha5gsm** [reply]
Thanks for all the the clarifications! This helps.
